src/libcore/str/mod.rs

```diff
@@ -1741,13 +1741,29 @@ fn truncate_to_char_boundary(s: &str, mut max: usize) -> (bool, &str) {
 #[cold]
 fn slice_error_fail(s: &str, begin: usize, end: usize) -> ! {
     const MAX_DISPLAY_LENGTH: usize = 256;
-    let (truncated, s) = truncate_to_char_boundary(s, MAX_DISPLAY_LENGTH);
+    let (truncated, s_trunc) = truncate_to_char_boundary(s, MAX_DISPLAY_LENGTH);
     let ellipsis = if truncated { "[...]" } else { "" };
 
+    // 1. out of bounds
+    let index = if !s.is_char_boundary(begin) { begin } else { end };
+    if begin > s.len() || end > s.len() {
+        panic!("index {} is out of bounds of `{}`{}", index, s_trunc, ellipsis);
+    }
+
+    // 2. begin <= end
     assert!(begin <= end, "begin <= end ({} <= {}) when slicing `{}`{}",
-            begin, end, s, ellipsis);
-    panic!("index {} and/or {} in `{}`{} do not lie on character boundary",
-          begin, end, s, ellipsis);
+            begin, end, s_trunc, ellipsis);
+
+    // 3. character boundary
+    // find the character
+    let mut char_start = index;
+    while !s.is_char_boundary(char_start) {
+        char_start -= 1;
+    }
+    let ch = s[char_start..].chars().next().unwrap();
```

Comment:
This can panic if the string is empty or if `char_start` is at the end. I think I've convinced myself this can't happen, but at least a comment would help, or just to be defensive

```rust
if let Some(ch) = s[char_start..].chars().next() {
    let char_range = char_start .. char_start + ch.len_utf8();
    panic!("index {} is not a char boundary; it is inside {:?} (bytes {:?}) of `{}`{}",
           index, ch, char_range, s_trunc, ellipsis);
} else {
    panic!("index {} is not a char boundary in `{}`{}", index, s_trunc, ellipsis);
}
```